rectools/models/ease.py

```diff
@@ -0,0 +1,134 @@
+#  limitations under the License.
+
+"""EASE model."""
+
+import typing as tp
+
+import numpy as np
+from scipy import sparse
+from tqdm import tqdm
+
+from rectools import InternalIds
+from rectools.dataset import Dataset
+
+from .base import ModelBase, Scores
+from .utils import recommend_from_scores
+from .vector import Distance, ImplicitRanker
+
+
+class EASEModel(ModelBase):
+    """
+    Embarrassingly Shallow Autoencoders for Sparse Data model.
+
+    See https://arxiv.org/abs/1905.03375.
+
+    Parameters
+    ----------
+    weight : np.matrix
+        The Item-Item weight-matrix.
+    verbose : int, default 0
+        Degree of verbose output. If 0, no output will be provided.
+    """
+
+    u2i_dist = Distance.DOT
+
+    def __init__(
+        self,
+        regularization: float = 500.0,
+        verbose: int = 0,
+    ):
+        super().__init__(verbose=verbose)
+        self.regularization = regularization
+        self.weight: np.ndarray
+
+    def _fit(self, dataset: Dataset) -> None:  # type: ignore
+        ui_csr = dataset.get_user_item_matrix(include_weights=True)
+
+        gram_matrix = ui_csr.T @ ui_csr
+        gram_matrix += self.regularization * sparse.identity(gram_matrix.shape[0]).astype(np.float32)
+        gram_matrix = gram_matrix.todense()
+
+        gram_matrix_inv = np.linalg.inv(gram_matrix)
+
+        self.weight = np.array(gram_matrix_inv / (-np.diag(gram_matrix_inv)))
+        np.fill_diagonal(self.weight, 0.0)
+
+    def _recommend_u2i(
+        self,
+        user_ids: np.ndarray,
+        dataset: Dataset,
+        k: int,
+        filter_viewed: bool,
+        sorted_item_ids_to_recommend: tp.Optional[np.ndarray],
+    ) -> tp.Tuple[InternalIds, InternalIds, Scores]:
+        user_items = dataset.get_user_item_matrix(include_weights=True)
+
+        ranker = ImplicitRanker(
+            distance=self.u2i_dist,
+            subjects_factors=user_items,
+            objects_factors=self.weight,
+        )
+        ui_csr_for_filter = user_items[user_ids] if filter_viewed else None
+
+        all_user_ids, all_reco_ids, all_scores = ranker.rank(
+            subject_ids=user_ids,
+            k=k,
+            filter_pairs_csr=ui_csr_for_filter,
+            sorted_object_whitelist=sorted_item_ids_to_recommend,
+            num_threads=0,
+        )
+
+        return all_user_ids, all_reco_ids, all_scores
+
+    def _recommend_i2i(
+        self,
+        target_ids: np.ndarray,
+        dataset: Dataset,
+        k: int,
+        sorted_item_ids_to_recommend: tp.Optional[np.ndarray],
+    ) -> tp.Tuple[InternalIds, InternalIds, Scores]:
+        similarity = self.weight
+        if sorted_item_ids_to_recommend is not None:
+            similarity = similarity[:, sorted_item_ids_to_recommend]
+
+        all_target_ids = []
+        all_reco_ids: tp.List[np.ndarray] = []
+        all_scores: tp.List[np.ndarray] = []
+        for target_id in tqdm(target_ids, disable=self.verbose == 0):
```

Comment:
Cycles of this type are very slow for inference. We do this only when we have no other option. But here you have a very nice filled item-item similarity matrix in np.array format.

If we have similarity scores array for one item and need recommendations for it, we can do:
```
unsorted_reco_positions = scores.argpartition(-k)[-k:]
unsorted_reco_scores = scores[unsorted_reco_positions]
order = unsorted_reco_scores.argsort()[::-1]
reco = unsorted_reco_positions[order]
scores = unsorted_reco_scores[order]
```

But in EASE recommend we have scores as a matrix and need to process recommendations for a couple of items. We can do smth like:

```
unsorted_reco_positions = scores.argpartition(-k, axis=1)[:, -k:]
unsorted_reco_scores = np.take_along_axis(scores, unsorted_reco_positions, axis=1)
reco = np.take_along_axis(
            unsorted_reco_positions, unsorted_reco_scores.argsort()[:, ::-1], axis=1
        )
```

Reply:
We should probably add batches processing for this logic.

Another way is to use ImplicitRanker again. Just feed identity matrix as subject factors. It could be faster this way due to implicit library realisation.

Reply:
It might be worth changing the `recommend_from_scores` function so that it works with scores with dimensions `(1, n_items)` rather than `(n_items,)`?